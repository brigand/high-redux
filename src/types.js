// @flow
'use strict';

import type HrQuery from './HrQuery';

export function getKey(key: ?string) {
  if (!key) return '[[default]]';
  return key;
}

export type HrStateDesc<T> = {
  loading: boolean,
  hasError: boolean,
  error: ?any,
  value: T,
  loadingStartTime: number,
  loadingCompleteTime: number,
  etc: Object,
}

export type PathType = 'id' | 'list' | 'kv';
export type OpType = 'setIds' | 'mergeIds' | 'updateInDesc' | 'deleteDesc' | 'listOp' | 'clearOptimistic' | 'mergeDesc';

export type HrStateWrapperOp = {
  op: OpType,
  data: any,
  type: ?PathType,
  typeValue: string | null,
  key: ?string,
  optimisticId: ?string,
};

export type HrStateById = { [type: string]: { [id: string]: HrStateDesc<any> } };
export type HrStateList = { [type: string]: HrStateDesc<Array<any>> };
export type HrStateKv = { [type: string]: { [id: string]: HrStateDesc<any> } };
export type RollbackOp = HrStateWrapperOp;
export type RollbackOps = { [id: string]: Array<RollbackOp> };

export type HrState = {
  isHrState: true,
  byId: HrStateById,
  lists: HrStateList,
  kv: HrStateKv,
  rollbackOps: RollbackOps,
}


export type SelectorFunc = (query: HrQuery, state: Object) => any
export type SelectorMapping = { [selectorName: string]: SelectorFunc }

export type SubscribeDescriptor = {
  handler: (selfState: Object, ownProps: Object, dispatch: Function, state: ?Object) => mixed,
  needsState?: boolean,
}
