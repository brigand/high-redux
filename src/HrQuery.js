// @flow
import * as t from './types';

export default class HrQuery {
  st: Object;

  /*
    Create an HrQuery for the given object, which often implements the HrState
    interface. Most methods don't work on any other object.
  */
  constructor(stateTree: Object) {
    this.st = stateTree
  }

  /*
    Get the value with the specified id, or null if it doesn't exist.
  */
  valueById(id: string) {
    return this.valueByIdKey(t.getKey(null), id);
  }
  valueByIdKey(key: string, id: string) {
    const desc = this.descByIdKey(key, id);
    return desc ? desc.value : null;
  }

  /*
    Get a record by id, and return an object with a set of props that can be
    passed to a React component or similar.

    For example, var p = propsById(someId); p.data is the data, and p.dataLoading
    is the loading boolean.

    A name can be specified, which will be the prefix of the prop names.

    var p = propsById(someId, 'foo') gives p.foo as the main data, and p.fooLoading
    as the loading boolean.
  */
  propsById(id: string, name: ?string = null) {
    return this.propsByIdKey(t.getKey(null), id, name);
  }
  propsByIdKey(key: string, id: string, _name: ?string = null) {
    const name = _name || 'data';

    const desc = this.descByIdKey(key, id);
    if (!desc) return {};

    return {
      [name]: desc.value,
      [`${name}Error`]: desc.error,
      [`${name}HasError`]: desc.hasError,
      [`${name}Loading`]: desc.loading,
      [`${name}Meta`]: desc.userMeta || {},
    };
  }

  /*
    Get the default list. Always returns an array, but it may be empty
  */
  list() {
    return this.listKey(t.getKey(null));
  }
  listKey(key: string) {
    const obj = this._assertHr(`list`);
    return obj.lists[key] && obj.lists[key].value || [];
  }

  /*
    Get props for the list state. See the propsById docs for an example
    of how this works.
  */
  propsFromList(name: ?string = null) {
    return this.propsFromListKey(t.getKey(null), name);
  }
  propsFromListKey(key: string, _name: ?string = null) {
    const name = _name || 'data';

    const desc = this.descListKey(key);
    if (!desc) return {};

    return {
      [name]: desc.value,
      [`${name}Error`]: desc.error,
      [`${name}HasError`]: desc.hasError,
      [`${name}Loading`]: desc.loading,
      [`${name}Meta`]: desc.userMeta || {},
    };
  }

  /*
    Get the state descriptor for the given id. Mostly for internal use.
  */
  descById(id: string) {
    return this.descByIdKey(t.getKey(null), id);
  }
  descByIdKey(key: string, id: string) {
    const obj = this._assertHr(`descByIdKey`);
    const desc = obj.byId[key][id];
    return desc || null;
  }

  /*
    Get the state descriptor for the list. Mostly for internal use.
  */
  descList() {
    return this.descListKey(t.getKey(null));
  }
  descListKey(key: string) {
    const obj = this._assertHr(`descListKey`);
    const desc = obj.lists[key];
    return desc || null;
  }

  /*
    Get the state descriptor for the given key/value pair. Mostly for internal use.
  */
  descKv(id: string) {
    return this.descKvKey(t.getKey(null), id);
  }
  descKvKey(key: string, id: string) {
    const obj = this._assertHr(`descByIdKey`);
    const desc = obj.kv[key][id];
    return desc || null;
  }

  /*
    Asserts that we have an HR state object. For internal use.
  */
  _assertHr(method: string, obj: Object = this.st): t.HrState {
    if (!obj.isHrState) {
      throw new Error(`HrQuery::${method} requires a high-redux state object but got {${Object.keys(obj).join(', ')}}`);
    }
    return obj;
  }
}
