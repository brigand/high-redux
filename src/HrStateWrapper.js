// @flow
'use strict';
import * as t from './types';
import HrQuery from './HrQuery';

// eslint-disable-next-line

export function makeDefaultHrState(): t.HrState {
  return {
    isHrState: true,
    byId: {},
    lists: {},
    kv: {},
    rollbackOps: {},
  };
}

export function makeHrStateDesc<T>(value: T, properties: ?$Shape<t.HrStateDesc<T>>): t.HrStateDesc<T> {
  return {
    loading: false,
    hasError: false,
    error: null,
    value,
    loadingStartTime: 0,
    loadingCompleteTime: 0,
    etc: {},
    ...properties,
  };
}

type StateInfo = {
  state: t.HrState,
  ops: Array<t.HrStateWrapperOp>,
  time: number,
}

type StatePath = {
  isHrStatePath: true,

  // The type of selector, defaults to null.
  type: ?t.PathType,

  // The value for the selector. Used by id and kv types
  typeValue: string | null,

  // The state key, which defaults to null
  key: string | null,

  info: StateInfo,
  parent: ?HrStateWrapper,

  optimisticId: ?string,
}

const TEST_TIME = 1500000000000;

function makeInfo(state: t.HrState): StateInfo {
  const infoObj = {
    state,
    ops: [],
    time: process.env.NODE_ENV === 'test' ? TEST_TIME : Date.now(),
  };

  return infoObj;
}

function makePath(info: StateInfo, parent: ?HrStateWrapper): StatePath {
  const pathObj = {
    isHrStatePath: true,
    type: parent ? parent.path.type : null,
    typeValue: parent ? parent.path.typeValue : null,
    key: parent ? parent.path.key : null,
    info,
    parent,
    optimisticId: parent ? parent.path.optimisticId : null,
  };

  return pathObj;
}

export function wrapperFromState(_state: ?t.HrState) {
  const state: t.HrState = _state || makeDefaultHrState();
  if (process.env.NODE_ENV === 'test') require('deep-freeze')(state);

  const info = makeInfo(state);
  const path = makePath(info);
  return new HrStateWrapper(path);
}

export class HrStateWrapper {
  path: StatePath
  /*
    Creates a HrStateWrapper instance. This is typically done for you.
  */
  constructor(path: StatePath) {
    this.path = path;
  }

  /*
    Utility for running a function without breaking the chain. Receives the
    `HrStateWrapper` as an argument.
  */
  invoke(fn: (s: HrStateWrapper) => mixed) {
    fn(this);
    return this;
  }

  /*
    Returns a state wrapper for the specified `id`. You can then call methods
    like `.set` on it.

    ```javascript
    s.id('some-id').set('some-value')
    ```
  */
  id(id: string) {
    const path = makePath(this.path.info, this);
    path.type = 'id';
    path.typeValue = id;
    return new HrStateWrapper(path);
  }

  /*
    Returns a state wrapper for the list.

    ```javascript
    s.list().set(['a', 'b'])
    ```
  */
  list() {
    const path = makePath(this.path.info, this);
    path.type = 'list';
    path.typeValue = null;
    return new HrStateWrapper(path);
  }

  /*
    Returns a state wrapper for the specified key in the key/value pair.

    ```javascript
    s.kv('some-key').set('some-value')
    ```
  */
  kv(kvKey: string) {
    const path = makePath(this.path.info, this);
    path.type = 'kv';
    path.typeValue = kvKey;
    return new HrStateWrapper(path);
  }

  /*
    Returns a state wrapper for the specified sub-key.

    ```javascript
    s.key('some-key').id('some-id').set('foo')
    ```
  */
  key(key: string) {
    const path = makePath(this.path.info, this);
    path.key = key;
    return new HrStateWrapper(path);
  }

  /*
    Get an `HrQuery` object for the state.
  */
  queryRoot() {
    return new HrQuery(this.path.info.state);
  }

  query() {
    let q = new HrQuery(this.path.info.state);
    if (this.path.key) q = q.key(this.path.key);

    if (this.path.type === 'id' && this.path.typeValue) {
      q = q.id(this.path.typeValue);
    } else if (this.path.type === 'list') {
      q = q.list();
    } else if (this.path.type === 'kv' && this.path.typeValue) {
      q = q.kv(this.path.typeValue);
    }

    return q;
  }

  /*
    Sets the currently focused item. See the examples for `id`/`list`/`kv`
  */
  set(value: any) {
    this._pushOp('updateInDesc', {
      path: ['value'],
      value,
      merge: false,
    });
    return this;
  }

  /*
    For each item in `pairs`, map the first item in the pair (the id) to the
    second item (the value).

    Often you'll generate this with an `array.map` call.

    ```javascript
    s.setIdPairs(items.map(x => [x.id, x]))
    s.key('some-key').setIdPairs(items.map(x => [x.id, x]))
    ```
  */
  setIds(pairs: Array<[string, any]>) {
    this._pushOp('setIds', pairs, { type: 'id' });
    return this;
  }

  /*
    Update an item with the given id by passing it to the 'updater' function.
  */
  update(updater: Function) {
    this._pushOp('updateInDesc', { path: ['value'], updater, merge: true });
    return this;
  }

  /*
    Update an item with the given id by passing it to the 'updater' function.
  */
  updateIn(updatePath: Array<string | number>, updater: Function) {
    this._pushOp('updateInDesc', { path: ['value'].concat(updatePath), updater, merge: true });
    return this;
  }

  /*
    Set the focused item to a loading state. Also captures the loading state.

    ```javascript
    s.id('some-id').setLoading()
    ```
  */
  setLoading() {
    this._pushOp('updateInDesc', {
      path: ['hasError'],
      value: false,
    });
    this._pushOp('updateInDesc', {
      path: ['loading'],
      value: true,
    });
    this._pushOp('updateInDesc', {
      path: ['loadingStartTime'],
      value: this.path.info.time,
    });

    return this;
  }

  /*
    Set the focused item to a completed loading state.

    ```javascript
    s.id('some-id').setLoadingDone()
    ```
  */
  setLoadingDone() {
    this._pushOp('updateInDesc', {
      path: ['hasError'],
      value: false,
    });
    this._pushOp('updateInDesc', {
      path: ['loading'],
      value: false,
    });
    this._pushOp('updateInDesc', {
      path: ['loadingCompleteTime'],
      value: this.path.info.time,
    });

    return this;
  }

  /*
    Set the focused item to an error state. Pass `null` to clear the error state.

    ```javascript
    s.id('some-id').setError({ message: 'Oops' })
    ```
  */
  setError(error: ?any) {
    this._pushOp('updateInDesc', {
      path: ['hasError'],
      value: error != null,
    });
    this._pushOp('updateInDesc', {
      path: ['error'],
      value: error,
    });
    this._pushOp('updateInDesc', {
      path: ['loading'],
      value: false,
    });
    this._pushOp('updateInDesc', {
      path: ['loadingCompleteTime'],
      value: this.path.info.time,
    });

    return this;
  }

  /*
    Set custom metadata for the current item.

    ```javascript
    s.id('some-id').setMeta('isNew', true)
    ```
  */
  setMeta(metaKey: string, metaValue: any) {
    this._pushOp('updateInDesc', { path: ['etc', metaKey], value: metaValue });

    return this;
  }

  /*
    Push an item to the list.

    Throws if not in list mode.
  */
  push(...items: Array<any>) {
    if (this.path.type !== 'list') {
      throw new Error(`Attempted to push while in ${this.path.type || '(none)'} mode.`);
    }
    this._pushOp('listOp', { items: items, listOp: 'push' });
    return this;
  }

  /*
    Adds an item to the start of the list.

    Throws if not in list mode.
  */
  unshift(...items: Array<any>) {
    if (this.path.type !== 'list') {
      throw new Error(`Attempted to push while in ${this.path.type || '(none)'} mode.`);
    }
    this._pushOp('listOp', { items: items, listOp: 'unshift' });

    return this;
  }

  /*
    Removes items from the end of the list. The count defaults to 1.

    Throws if not in list mode.
  */
  pop(count: number = 1) {
    if (this.path.type !== 'list') {
      throw new Error(`Attempted to push while in ${this.path.type || '(none)'} mode.`);
    }
    this._pushOp('listOp', { count, listOp: 'pop' });

    return this;
  }

  /*
    Removes items from the start of the list. The count defaults to 1.

    Throws if not in list mode.
  */
  shift(count: number = 1) {
    if (this.path.type !== 'list') {
      throw new Error(`Attempted to push while in ${this.path.type || '(none)'} mode.`);
    }
    this._pushOp('listOp', { count, listOp: 'shift' });

    return this;
  }

  /*
    Set this operation to be optimistic, which can be rolled back on future
    state wrappers.

    ```javascript
    s.optimistic('token').id('some-id').set(value);

    // in the future
    s.optimistic('token').rollback()
    ```
  */
  optimistic(id: ?string) {
    const id2 = id || '[[default]]';

    const path = makePath(this.path.info, this);
    path.optimisticId = id2;
    return new HrStateWrapper(path);
  }

  /*
    Rolls back a previous optimistic update.

    ```javascript
    s.optimistic('token').rollback()
    ```
  */
  rollback() {
    const id = this.path.optimisticId;

    if (!id) {
      throw new Error(`You must call .optimistic before you can rollback an update.`);
    }

    const ops = this.path.info.state.rollbackOps[id];
    if (!ops) {
      throw new Error(`Attempted to rollback optimistic update "${id}" which doesn't exist.`);
    }

    this.path.info.ops.push(...ops);

    return this;
  }

  /*
    Compute the state by applying all update operations. Mostly for internal use.
  */
  getState(): Object {
    const sw: HrStateWrapper = this.root();
    let { state, ops } = sw.path.info;
    if (!ops.length) return state;

    if (process.env.NODE_ENV === 'test') require('deep-freeze')(state);

    state = { ...state };

    // Track which things we've cloned to improve performance, and avoid
    // cloning objects that don't need to be cloned.
    const cloned = { id: false, list: false, kv: false, rollbackOps: false };
    const clonedKey = {
      id: Object.create(null),
      list: Object.create(null),
      kv: Object.create(null),
    };

    const updatedRollbacks = {};
    const addRollback = (id: string, op: t.HrStateWrapperOp) => {
      if (!cloned.rollbackOps) {
        cloned.rollbackOps = true;
        state.rollbackOps = { ...state.rollbackOps };
      }

      if (!updatedRollbacks[id]) {
        state.rollbackOps[id] = [];
        updatedRollbacks[id] = true;
      }

      op.optimisticId = null;

      state.rollbackOps[id].push(op);
    }

    for (let i = 0; i < ops.length; i += 1) {
      const op = ops[i];

      const key = t.getKey(op.key);

      if (op.type === 'id' || op.type === 'kv') {
        const stateKey = op.type === 'id' ? 'byId' : 'kv';
        if (!cloned[op.type]) {
          cloned[op.type] = true;
          state[stateKey] = { ...state[stateKey] };
        }

        if (!clonedKey[op.type][key]) {
          clonedKey[op.type][key] = true;
          state[stateKey][key] = { ...state[stateKey][key] };
        }

        if (op.op === 'set') {
          const opts: $Shape<t.HrStateDesc<any>> = {};
          opts.loadingCompleteTime = this.path.info.time;
          // $FlowFixMe
          if (state[stateKey][key][op.typeValue]) {
            opts.loadingStartTime = state[stateKey][key][op.typeValue].loadingStartTime;
          }

          if (op.optimisticId) {
            addRollback(op.optimisticId, {
              ...op,
              // $FlowFixMe
              data: state[stateKey][key][op.typeValue] !== undefined ? state[stateKey][key][op.typeValue].value : undefined,
            });
          }

          // $FlowFixMe
          state[stateKey][key][op.typeValue] = makeHrStateDesc(op.data, opts);
        }

        if (op.op === 'setIds') {
          const optimisticId = op.optimisticId;

          const newDescs = [];

          for (let i = 0; i < op.data.length; i += 1) {
            const pair = op.data[i];
            newDescs.push(makeHrStateDesc(pair[1]));
          }

          if (optimisticId) {
            const newDescsMapping = {};
            const oldDescPairs = [];
            const oldDescNotExist = [];
            for (let i = 0; i < op.data.length; i += 1) {
              const pair = op.data[i];
              const curr = state[stateKey][key][pair[0]];
              newDescsMapping[pair[0]] = newDescs[i];
              if (curr) {
                const oldValue = state[stateKey][key][pair[0]];
                oldDescPairs.push([pair[0], oldValue]);
              } else {
                oldDescNotExist.push(pair[0]);
              }
            }

            for (const pair of oldDescPairs) {
              addRollback(optimisticId, {
                ...op,
                type: 'id',
                typeValue: pair[0],
                op: 'updateInDesc',
                data: {
                  path: [],
                  value: pair[1],
                },
              });
            }

            for (const id of oldDescNotExist) {
              addRollback(optimisticId, {
                ...op,
                type: 'id',
                typeValue: id,
                op: 'deleteDesc',
                data: {},
              });
            }
          }

          for (let i = 0; i < op.data.length; i += 1) {
            const pair = op.data[i];
            state[stateKey][key][pair[0]] = newDescs[i];
          }
        }

        if (op.op === 'deleteDesc') {
          const { optimisticId } = op;

          // $FlowFixMe
          const current = state[stateKey][key][op.typeValue];

          if (current) {
            // $FlowFixMe
            delete state[stateKey][key][op.typeValue];

            if (optimisticId) {
              addRollback(optimisticId, {
                ...op,
                op: 'updateInDesc',
                data: {
                  path: ['value'],
                  data: current,
                },
              });
            }
          }
        }

        if (op.op === 'updateInDesc') {
          // $FlowFixMe
          const desc = state[stateKey][key][op.typeValue] ? { ...state[stateKey][key][op.typeValue] } : makeHrStateDesc(null);

          const opPath = [...op.data.path];
          let final = opPath.slice(0, -1).reduce((acc, k) => {
            desc[k] = { ...desc[k] };
            return desc[k];
          }, desc);

          // Apply the updater function to the new value
          const prop = opPath[opPath.length - 1];

          const updateValue = op.data.updater ? op.data.updater(prop ? final[prop] : final) : op.data.value;
          let newValue = updateValue;
          let changed = null;

          const { optimisticId } = op;

          // If it's a non-array object, then merge it into the previous value
          if (newValue && typeof newValue === 'object' && !Array.isArray(newValue)) {
            newValue = { ...final[prop], ...newValue };

            if (optimisticId) {
              const changed = Object.keys(updateValue);
              const original = changed.reduce((acc, key) => {
                if (final[prop]) {
                  acc[key] = final[prop][key];
                } else {
                  throw new Error(`Something is wrong in updateInDesc`);
                }
                return acc;
              }, {});

              addRollback(optimisticId, {
                ...op,
                data: {
                  path: op.data.path,
                  value: original,
                },
              });
            }

            if (prop) {
              final[prop] = newValue;
            } else {
              final = newValue;
            }
          } else {
            if (optimisticId) {
              addRollback(optimisticId, {
                ...op,
                data: {
                  path: op.data.path,
                  value: prop ? final[prop] : final,
                },
              });
            }
          }

          if (!prop) {
            // $FlowFixMe
            state[stateKey][key][op.typeValue] = final;
          } else {
            final[prop] = newValue;
            // $FlowFixMe
            state[stateKey][key][op.typeValue] = desc;
          }
        }
      }

      if (op.type === 'list') {
        if (!cloned.list) {
          cloned.list = true;
          state.lists = { ...state.lists };
        }

        if (op.op === 'updateInDesc') {
          const desc = state.lists[key] ? { ...state.lists[key] } : makeHrStateDesc(null);

          const opPath = [...op.data.path];

          let final = opPath.slice(0, -1).reduce((acc, k) => {
            desc[k] = { ...desc[k] };
            return desc[k];
          }, desc);

          // Apply the updater function to the new value
          const prop = opPath[opPath.length - 1];

          const updateValue = op.data.updater ? op.data.updater(prop ? final[prop] : final) : op.data.value;
          let newValue = updateValue;

          const { optimisticId } = op;

          if (optimisticId) {
            addRollback(optimisticId, {
              ...op,
              data: {
                path: op.data.path,
                value: prop ? final[prop] : final,
              },
            });
          }

          if (prop) {
            final[prop] = newValue;
          } else {
            final = newValue;
          }

          // $FlowFixMe
          state.lists[key] = desc;
        }

        if (op.op === 'listOp') {
          const { optimisticId } = op;
          const { listOp, count, items } = op.data;
          const desc = state.lists[key] ? { ...state.lists[key] } : makeHrStateDesc([]);

          if (listOp === 'push') {
            desc.value = desc.value.concat(items);
            state.lists[key] = desc;

            if (optimisticId) {
              addRollback(optimisticId, {
                ...op,
                op: 'listOp',
                data: {
                  listOp: 'removeItems',
                  items,
                },
              });
            }
          }
          if (listOp === 'unshift') {
            desc.value = items.concat(desc.value);
            state.lists[key] = desc;

            if (optimisticId) {
              addRollback(optimisticId, {
                ...op,
                op: 'listOp',
                data: {
                  listOp: 'removeItems',
                  items,
                },
              });
            }
          }
          if (listOp === 'pop') {
            const current = desc.value || [];

            const removed = current.slice(-1 * count);
            desc.value = current.slice(0, -1 * count);
            state.lists[key] = desc;

            if (optimisticId) {
              addRollback(optimisticId, {
                ...op,
                op: 'listOp',
                data: {
                  listOp: 'push',
                  items: removed,
                },
              });
            }
          }

          if (listOp === 'shift') {
            const current = desc.value || [];

            const removed = current.slice(0, count);
            desc.value = current.slice(count);
            state.lists[key] = desc;

            if (optimisticId) {
              addRollback(optimisticId, {
                ...op,
                op: 'listOp',
                data: {
                  listOp: 'unshift',
                  items: removed,
                },
              });
            }
          }
          if (listOp === 'removeItems') {
            desc.value = desc.value.filter(x => items.indexOf(x) === -1);
            state.lists[key] = desc;
          }
        }
      }
    }

    return state;
  }

  /*
    Get the root `HrStateWrapper` instance.
  */
  root() {
    let wrapper: HrStateWrapper = this;
    while (wrapper.path.parent) {
      wrapper = wrapper.path.parent;
    }

    return wrapper;
  }

  /*
    Internal: adds an operation to the queue
  */
  _pushOp(op: t.OpType, data: any, overrides: $Shape<t.HrStateWrapperOp> = {}) {
    this.path.info.ops.push({
      op,
      key: this.path.key,
      type: this.path.type,
      typeValue: this.path.typeValue,
      data,
      optimisticId: this.path.optimisticId,
      ...overrides,
    });
  }
}
