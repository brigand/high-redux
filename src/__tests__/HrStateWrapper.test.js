import { wrapperFromState } from '../HrStateWrapper';
import HrQuery from '../HrQuery';
import { getKey } from '../types';

describe(`HrStateWrapper basic`, () => {
  it(`doesn't error`, () => {
    const s = wrapperFromState();
    s.setIds([['some-id', 'some-value']]);
    s.id('another-id').set('another-value');
    s.id('another-id').setError({ message: 'test error' });
    s.list().set(['some-id', 'another-id']);
    const state = s.getState();

    expect(state).toBeInstanceOf(Object);
    expect(state).toMatchSnapshot(`doesn't error: testing many operations`);
  });
});

describe(`HrStateWrapper id`, () => {
  it(`setIds`, () => {
    const s = wrapperFromState();
    s.setIds([['idA', 'valueA'], ['idB', 'valueB']]);

    const state = s.getState();
    expect(state.byId[getKey()].idA.value).toBe('valueA');
    expect(state.byId[getKey()].idB.value).toBe('valueB');
  });

  it(`id->set`, () => {
    const s = wrapperFromState();
    s.setIds([['idA', 'valueA'], ['idB', 'valueB']]);
    s.id('idC').set('valueC');

    const state = s.getState();
    expect(state.byId[getKey()].idA.value).toBe('valueA');
    expect(state.byId[getKey()].idB.value).toBe('valueB');
    expect(state.byId[getKey()].idC.value).toBe('valueC');
  });

  it(`id->setError`, () => {
    const s = wrapperFromState();

    s.id('idA').set('valueA');
    s.id('idA').setError('some-error');
    s.id('idB').setError('another-error');

    const state = s.getState();
    expect(state.byId[getKey()].idA.value).toBe('valueA');
    expect(state.byId[getKey()].idA.hasError).toBe(true);
    expect(state.byId[getKey()].idB.hasError).toBe(true);

    expect(state.byId[getKey()].idA.error).toBe('some-error');
    expect(state.byId[getKey()].idB.error).toBe('another-error');
  });

  it(`alternate keys`, () => {
    const s = wrapperFromState();
    s.key('foo').id('id').set('valueA');
    s.id('id').set('valueB');
    s.id('id').set('valueB');
    s.key('bar').id('id').set('valueC');
    s.key('bar').id('id2').set('valueD');

    const state = s.getState();
    expect(state.byId[getKey('foo')].id.value).toBe('valueA');
    expect(state.byId[getKey(null)].id.value).toBe('valueB');
    expect(state.byId[getKey('bar')].id.value).toBe('valueC');
    expect(state.byId[getKey('bar')].id2.value).toBe('valueD');
  });
});

describe(`HrStateWrapper list`, () => {
  it(`setList`, () => {
    const s = wrapperFromState();
    s.list().set(['v1', 'v2']);

    const state = s.getState();
    expect(state.lists[getKey()].value).toEqual(['v1', 'v2']);
  });

  it(`push`, () => {
    const s = wrapperFromState().list().set(['v1', 'v2']);

    const s2 = wrapperFromState(s.getState()).optimistic('foo').list().push('v3', 'v4');

    expect(s2.getState().lists[getKey()].value).toEqual(['v1', 'v2', 'v3', 'v4']);

    const s3 = wrapperFromState(s2.getState()).optimistic('foo').rollback();
    expect(s3.getState().lists[getKey()].value).toEqual(['v1', 'v2']);
  });

  it(`unshift`, () => {
    const s = wrapperFromState().list().set(['v1', 'v2']);

    const s2 = wrapperFromState(s.getState()).optimistic('foo').list().unshift('v3', 'v4');

    expect(s2.getState().lists[getKey()].value).toEqual(['v3', 'v4', 'v1', 'v2']);

    const s3 = wrapperFromState(s2.getState()).optimistic('foo').rollback();
    expect(s3.getState().lists[getKey()].value).toEqual(['v1', 'v2']);
  });

  it(`pop`, () => {
    const s = wrapperFromState().list().set(['v1', 'v2', 'v3', 'v4']);

    const s2 = wrapperFromState(s.getState()).optimistic('foo').list().pop(2);

    expect(s2.getState().lists[getKey()].value).toEqual(['v1', 'v2']);

    const s3 = wrapperFromState(s2.getState()).optimistic('foo').rollback();
    expect(s3.getState().lists[getKey()].value).toEqual(['v1', 'v2', 'v3', 'v4']);
  });

  it(`shift`, () => {
    const s = wrapperFromState().list().set(['v1', 'v2', 'v3', 'v4']);

    const s2 = wrapperFromState(s.getState()).optimistic('foo').list().shift(2);

    expect(s2.getState().lists[getKey()].value).toEqual(['v3', 'v4']);

    const s3 = wrapperFromState(s2.getState()).optimistic('foo').rollback();
    expect(s3.getState().lists[getKey()].value).toEqual(['v1', 'v2', 'v3', 'v4']);
  });
});


describe(`HrStateWrapper kv`, () => {
  it(`setKv`, () => {
    const s = wrapperFromState();
    s.kv('test').set('val');

    const state = s.getState();
    expect(state.kv[getKey()].test.value).toBe('val');
    expect(state.kv[getKey()].test.hasError).toBe(false);
  });

  it(`setKvMeta`, () => {
    const s = wrapperFromState();
    s.kv('test').set('val');
    s.kv('test').setMeta('x', 1);

    const state = s.getState();
    expect(state.kv[getKey()].test.value).toBe('val');
    expect(state.kv[getKey()].test.etc.x).toBe(1);
  });
});

describe('HrStateWrapper optimistic', () => {
  it(`id/kv set`, () => {
    const s = wrapperFromState()
      .kv('foo').set('a').getState();

    const s2 = wrapperFromState(s).optimistic('op').kv('foo').set('b').getState();

    const s3 = wrapperFromState(s2).optimistic('op').rollback().getState();

    const q = new HrQuery(s3);
    expect(q.kv('foo')).toBe('a');
  });

  it(`id/kv updateValue`, () => {
    const s = wrapperFromState()
      .kv('foo').set({ a: 1, b: 2, c: 3 }).getState();

    const s2 = wrapperFromState(s).optimistic('op').kv('foo').update(o => ({
      b: 'this will be reverted',
    })).getState();

    const s3 = wrapperFromState(s2).kv('foo').update(o => ({ c: 3 })).getState();
    const s4 = wrapperFromState(s2).optimistic('op').rollback().getState();

    const q = new HrQuery(s4);
    expect(q.kv('foo')).toEqual({ a: 1, b: 2, c: 3 });
  });

  it(`id/kv setIds`, () => {
    const s = wrapperFromState()
      .setIds([
        ['ak', 'av'],
        ['bk', 'bv'],
      ]).getState();

    const s2 = wrapperFromState(s).optimistic('op').setIds([
      ['bk', 'this will be reset'],
      ['ck', 'this will be removed'],
    ]).getState();

    const s3 = wrapperFromState(s2).optimistic('op').rollback().getState();

    const q = new HrQuery(s3);
    expect(q.id('ak')).toBe('av');
    expect(q.id('bk')).toBe('bv');

    // Assert that we deleted the 'ck' key
    expect(s3.byId['[[default]]'].ak).toBeTruthy();
    expect(s3.byId['[[default]]'].ck).toBe(undefined);
  });

  it(`id/kv set new`, () => {
    const s = wrapperFromState()
      .id('foo').optimistic('op').set({ x: 1 }).getState();

    const q = new HrQuery(s);
    expect(q.id('foo')).toEqual({ x: 1 });

    const s2 = wrapperFromState(s).optimistic('op').rollback().getState();

    expect(s2.byId[getKey()]).toEqual({});
  });

  it(`clearOptimistic`, () => {
    const s = wrapperFromState()
      .kv('foo').set('a').getState();

    const s2 = wrapperFromState(s).optimistic('op').kv('foo').set('b').getState();

    const s3 = wrapperFromState(s2).clearOptimistic('op').getState();
    expect(s3.rollbackOps).toEqual({});
  });
});
