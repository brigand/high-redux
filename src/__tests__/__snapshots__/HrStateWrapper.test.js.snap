// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HrStateWrapper basic doesn't error: doesn't error: testing many operations 1`] = `
Object {
  "byId": Object {
    "[[default]]": Object {
      "another-id": Object {
        "error": Object {
          "message": "test error",
        },
        "hasError": true,
        "loading": false,
        "loadingCompleteTime": 0,
        "loadingStartTime": 0,
        "etc": Object {},
        "value": "another-value",
      },
      "some-id": Object {
        "error": null,
        "hasError": false,
        "loading": false,
        "loadingCompleteTime": 1514862765212,
        "loadingStartTime": 0,
        "etc": Object {},
        "value": "some-value",
      },
    },
  },
  "isHrState": true,
  "kv": Object {},
  "lists": Object {
    "[[default]]": Object {
      "error": null,
      "hasError": false,
      "loading": false,
      "loadingCompleteTime": 1514862765212,
      "loadingStartTime": 0,
      "etc": Object {},
      "value": Array [
        "some-id",
        "another-id",
      ],
    },
  },
}
`;
